src/main/java/org/zeromq/ZContext.java

```diff
@@ -116,10 +145,14 @@ public void destroySocket(Socket s)
         if (s == null) {
             return;
         }
-
-        if (sockets.remove(s)) {
-            s.setLinger(linger);
-            s.close();
+        s.setLinger(linger);
+        s.close();
+        try {
+            mutex.lock();
+            this.sockets.remove(s);
+        }
+        finally {
+            mutex.unlock();
```

Comment:
This part looks especially good to me.

The way the code is now, it seems like it's possible for a socket to sometimes not be closed, e.g. if it wasn't able to be removed because maybe it was removed in another thread.

The way you have it, we definitely close the socket, and then we remove it (within a locked mutex context, which is good), which seems safer.